### ISSUE TITLE
Implementing Sink<T> for Sender

### ISSUE THREAD

**Issue by Pantamis** [asynchronics/tachyonix#7]
Hello !

First thank you for your amazing work, your crates are amazing !

I would like to be able to implement the [`Sink` trait](https://docs.rs/futures/latest/futures/sink/trait.Sink.html) for the `Sender` endpoint of the channel.

Contrary to `Stream` for the `Receiver` the `Sink` trait is not in `futures-core` crate so I don't think it is worth implementing it in your crate to keep the dependancies minimal.

That's said it would be nice if your crate provides the needed methods to allow implementing `Sink<T>` in a clean way on a `Sender` wrapper. I think `try_send` already allow to implement the `start_send` method of the trait. What miss currently is buffering management, there is no public API to know if the queue is full to apply back-pressure (`poll_ready` method) or empty to let it flush (`poll_flush`) and maybe closing would be cleaner if we can use the `sender_signal` in the channel rather than creating a new `Event` to implement `poll_close` using `close` and `is_closed` and your `async-event` crate.

Otherwise I would need to include an atomic int and another sender signal in the `Sender` wrapper to track the number of items in the queue in the channel.

Do you think it is acceptable to provide such API in your crate ? It could be interesting to look at [how `tokio` does it for its channel](https://docs.rs/tokio-util/latest/src/tokio_util/sync/mpsc.rs.html#306-325), they do handle backpressure but not flushing.

Thank you for your time :)

**Comment by Pantamis** [comment]
After trying a bit it seems impossible to me to implement the sink correctly with a wrapper. So I gave a try at what can be done if I can touch the internals. I end up with:

Adding this method on the queue:

```rust
impl<T> Queue<T> {
    /// Perform checks to determine if a push can be attempted.
    /// The next push is guaranteed to be successful
    /// only if no other sender are waiting to send
    pub(super) fn can_push(&self) -> bool {
        let mut enqueue_pos = self.enqueue_pos.load(Ordering::Acquire);

        loop {
            let slot = &self.buffer[enqueue_pos & self.right_mask];
            let stamp = slot.stamp.load(Ordering::Acquire);

            let stamp_delta = stamp.wrapping_sub(enqueue_pos) as isize;

            match stamp_delta.cmp(&0) {
                cmp::Ordering::Less => return false,
                cmp::Ordering::Equal => return true,
                cmp::Ordering::Greater => enqueue_pos = self.enqueue_pos.load(Ordering::Acquire),
            }
        }
    }
}
```

to implement this method on `Sender`:

```rust
    /// Wait until a push can be attempted
    pub async fn wait_until_can_push(&self) {
        self.inner
            .sender_signal
            .wait_until(|| self.inner.queue.can_push().then_some(()))
            .await
    }
 ```
 
 to implement `Sink<T>` like this:
 
 ```rust
 impl<T> Sink<T> for Sender<T> {
    type Error = TrySendError<T>;

    fn poll_ready(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Result<(), Self::Error>> {
        pin!(self.wait_until_can_push()).poll(cx).map(Ok)
    }

    fn start_send(self: Pin<&mut Self>, item: T) -> Result<(), Self::Error> {
        self.try_send(item)
    }

    fn poll_flush(
        self: Pin<&mut Self>,
        _cx: &mut Context<'_>,
    ) -> Poll<Result<(), Self::Error>> {
        // Flushing does not really matter for a channel as long as done when it is closing
        Poll::Ready(Ok(()))
    }

    fn poll_close(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Result<(), Self::Error>> {
        self.close();
        pin!(self
            .inner
            .sender_signal
            .wait_until(|| self.is_closed().then_some(Ok(()))))
        .poll(cx)
    }
}
```

I don't think that we can guarantee that `try_send` will be successful right after when `wait_until_can_push` returns. But I think that this guarantee holds for the `Sink<T>` contract because all messages sent using the sink api to the channel will have to be preceed by a `wait_until_can_push` check through a `poll_ready` call. That's said I may not be fully aware of the raicy conditions that may happen on the `Queue<T>` and be wrong... I have to try it.

If this is right It could be worth it to add it behind a feature flag to avoid adding [`futures-sink` dependancy](https://crates.io/crates/futures-sink) just to get the `Sink` trait for those who don't need it.

**Comment by sbarral** [comment]
Many thanks for looking into this! I admit I have never used the `Sink` trait and indeed, unlike `Stream` it does not look like it has much chances ending up into `std`, but I am not against including support under a feature flag.

I haven't looked at the details yet but your analysis and suggestions about what needs to be implemented look correct to me, with one large caveat: I'm afraid that even if all messages are sent using the sink API (`poll_ready` followed by `start_send`), then `start_send` may fail if used concurrently. This is because the `can_push` method would return `true` to all concurrent senders even if only one slot is available. The `tokio` and `futures` channels have a different design than `tachyonix` or `async_channel`, and this allows them to guaranty that a send will succeed.
While we could theoretically "book" a slot by incrementing `enqueue_pos`, if `start_send` is not called fast enough this would ultimately block the channel once the enqueue position wraps around. TBH I don't really see a good solution here...

**Comment by Pantamis** [comment]
> This is because the can_push method would return true to all concurrent senders even if only one slot is available.

I think this is really the pain point.  A `Sink<T>` implementation would be sound if and only if `Sender` is not `Clone` for such channel. This would prevent them to race each others between a `poll_ready` and a `start_send`. In my use case I do not clone the sender so it would make sense... maybe 😅 ?

It could make sense to add the `Sink<T>` implementation behind a feature flag that also remove the `Clone` implementation and the related `sender_count` but I don't know if there is more efficient implementation of SPSC channel (most of them are oneshot) I should start to look at that first.

Thank you very much for taking time on this, interesting to see the tradeoffs like this one on concurrent programming.

**Comment by sbarral** [comment]
It does indeed look like your best bet is to use an SPSC channel, which would also be faster.

I haven't researched the topic but an obvious SPCS implementation would start with `tachyonix` and make `enqueue_pos` non-atomic (as `dequeue_pos` currently is). Then the enqueue operation simplifies a lot and becomes wait-free. When it comes to notifications, you could use `diatomic_waker` for both the sender and receiver, which is faster than `async_event`.

**Comment by Pantamis** [comment]
Quick update before closing:

For my use case I ended up "reversing the data flow logic" since my Sink is used to send data to a rayon thread pool and I mostly needed to get results back, I also think I needed some time to understand the intent behind the Sink abstraction. I successfully used your MPSC queue by arcing it. Then I needed to use the strong count of the Arc with atomic fences to monitor the number of handles to the queue to implement back pressure by forcing this count to be lower than a capacity.

It was working but I find it a bit too heavy to copy the `queue.rs` file with all the `unsafe` and atomic ordering choices it implies in my production code. So I end up refactoring my sink by using [oneshot](https://docs.rs/oneshot/latest/oneshot/index.html) channels (this implementation seems very light and efficient) which are created at `poll_ready` step if they are fewer then a capacity (a `diatomic-waker` is used to register and notifies this) and the sender is consumed when calling `start_send` as argument of a closure executed in rayon which reproduces a lighter system of permits as the tokio channel does. I am fine with allocating at this step since the thread pool is often already working in parallel while preparing the sink.

To received data back, the oneshot receivers are queued in a `FuturesUnorderedBounded` from [`futures-buffered`](https://docs.rs/futures-buffered/0.2.8/futures_buffered/) crate, which is also very light, uses the same MPSC queue from 1024cores, uses your amazing `diatomic-waker` (this one is really a gem) and is not using ugly unbounded linked lists as the `futures` crate. The `poll_flush` polls this queue to see if it can get going. 

So in the end I don't use `tachyonix` in my Sink at all and I don't think it is relevant for your crate anymore since it seems to me it cannot easily create "owned permits". But I learnd a lot in the process and I discover your work that I will follow closely, it is very very exciting thank you very much for it !

Cheers :)

**Comment by sbarral** [comment]
Be careful: concurrency primitives are a very deep rabbit hole ;-)

Anyway, glad you got it to work. I don't know if this can help but I wrote at some point a [`multishot`](https://docs.rs/multishot/latest/multishot/) crate which is a re-usable one-shot channel, with the upshot that allocation is done only once. Not sure it is pertinent to your use-case though, I admit I haven't fully grasped all subtleties...
Thanks about the pointer to these other crates, `futures-buffered` looks indeed quite interesting. I wrote as well a lighter variation of `FuturesUnordered` (`unsafe`-free!) but it is too idiosyncratic to make it into a general-purpose crate.
